pkg/controller/common/service_control.go

```diff
@@ -94,15 +93,15 @@ func applyServerSideValues(expected, reconciled *corev1.Service) {
 	// but might have been set after creation by k8s on the actual resource.
 	// In such case, we want to use these values for comparison.
 	// But only if we are not changing the type of service and the api server has assigned an IP
-	if expected.Spec.Type == reconciled.Spec.Type && expected.Spec.ClusterIP == "" && net.ParseIP(reconciled.Spec.ClusterIP) != nil {
+	if expected.Spec.Type == reconciled.Spec.Type && expected.Spec.ClusterIP == "" {
 		expected.Spec.ClusterIP = reconciled.Spec.ClusterIP
 	}
 
 	// ClusterIPs also might not exist in the expected service,
 	// but might have been set after creation by k8s on the actual resource.
 	// In such case, we want to use these values for comparison.
 	// But only if we are not changing the type of service and the api server has assigned IPs
-	if expected.Spec.Type == reconciled.Spec.Type && len(expected.Spec.ClusterIPs) == 0 && validClusterIPs(reconciled.Spec.ClusterIPs) {
+	if expected.Spec.Type == reconciled.Spec.Type && len(expected.Spec.ClusterIPs) == 0 {
 		expected.Spec.ClusterIPs = reconciled.Spec.ClusterIPs
 	}
```

Comment:
Nit: doesn't it make you want to combine the two?
```diff
diff --git a/pkg/controller/common/service_control.go b/pkg/controller/common/service_control.go
index 272ea82b0..f4ec0229e 100644
--- a/pkg/controller/common/service_control.go
+++ b/pkg/controller/common/service_control.go
@@ -89,20 +89,16 @@ func applyServerSideValues(expected, reconciled *corev1.Service) {
        if expected.Spec.Type == "" {
                expected.Spec.Type = reconciled.Spec.Type
        }
-       // ClusterIP might not exist in the expected service,
-       // but might have been set after creation by k8s on the actual resource.
-       // In such case, we want to use these values for comparison.
-       // But only if we are not changing the type of service and the api server has assigned an IP
-       if expected.Spec.Type == reconciled.Spec.Type && expected.Spec.ClusterIP == "" {
-               expected.Spec.ClusterIP = reconciled.Spec.ClusterIP
-       }
 
-       // ClusterIPs also might not exist in the expected service,
-       // but might have been set after creation by k8s on the actual resource.
-       // In such case, we want to use these values for comparison.
-       // But only if we are not changing the type of service and the api server has assigned IPs
-       if expected.Spec.Type == reconciled.Spec.Type && len(expected.Spec.ClusterIPs) == 0 {
-               expected.Spec.ClusterIPs = reconciled.Spec.ClusterIPs
+       // ClusterIP[s] might not exist in the expected service, but might have been set after creation by k8s on the actual resource.
+       // Use these values  for comparison, but only if we are not changing the type of service and the api server has assigned IP[s].
+       if expected.Spec.Type == reconciled.Spec.Type {
+               if expected.Spec.ClusterIP == "" {
+                       expected.Spec.ClusterIP = reconciled.Spec.ClusterIP
+               }
+               if len(expected.Spec.ClusterIPs) == 0 {
+                       expected.Spec.ClusterIPs = reconciled.Spec.ClusterIPs
+               }
        }
```

Reply:
Good point